### PR TITLE
Add continuous batching support for text-only models

### PR DESCRIPTION
**Pull request by neilmehta24**
The following PR description is long but it's not AI generated. Please read the sections you are interested in.

## Executive Summary

This PR adds continuous batching support to mlx-engine, enabling multiple concurrent generation requests to be processed in parallel on a single model. The implementation auto-detects batchable models and falls back to sequential generation when batching isn't supported.

## New Batched Generation Backend

`mlx-lm` provides a `BatchGenerator` class that provides 3 functions: `insert` `next` `remove`. The new backend instantiates `BatchGenerator` on a dedicated thread and runs in a loop. The loop picks up new requests and calls `insert` and picks up cancellation requests and calls `remove`. When there are no new requests, it runs `next` to generate a token per request. Up to `max_seq_nums` tokens will be generated.

`BatchGenerator` will collect requests that need to be prefilled. By default, `mlx-lm` chooses to wait for `8` requests that need prefill before stopping existing decodes. This means that one greedy decode sequence completely blocks all new prefill requests from being added to the decoding batch. So, reduce this batch size to `1`, so that all decoding requests will be halted by any prompt processing. This will make LM Studio more responsive, but it will also (oddly) pause all existing decode sequences. This is behavior that we would want to tune or configure down the road. An upstream implementation of chunked prefill could also help here.

The MLX framework is not (yet) threadsafe. This batched implementation was designed to ensure that all MLX operations are done on a single thread. Developers should not instantiate MLX arrays outside of this thread when using the batched backend.

To encapsulate the batched backend, create a new model kit called `BatchedModelKit`. This model kit will use the new batched generator defined in `generate.py`, and the existing model kits will use the existing generator (which is now renamed to sequential generator).

The prompt caching is managed with mlx-lm's LRU prompt cache. This enables (basic) automatic prefix caching. Note that the prefill batch size is increasing from 512 to 2048 as part of this change, in order to improve performance on more powerful machines. This means that some unit tests that test for the number of prompt processing callbacks were tweaked. The test input data file was also extended considering this new default size.

## New Threading Capabilities

Substantial additions were made to the existing model kits in order to enable multi-threading. The LM Studio mlx-engine driver took the Generator object returned from `create_generator`, and repeatedly called `__next__` on it to get the next token. This design meant that other work could be done between tokens calls, like registering a stop generation request.

With LM Studio 0.4, the engine driver changed to spawn a thread per prediction request. The stop generation request would also come in on a separate thread. This means that we must ensure that (1) we can concurrently call `__next__` on the batched Generator object, and (2) also forcibly serialize any requests that cannot be batched.

The BatchedModelKit relies on request ids to understand which requests should be stopped upon a cancel request. Each sequential request now requires a request id too. Each request can be cancelled at any point, whether it's waiting for the sequential generation lock, or if it's queued up to be batched.

The assumption is that each new thread that calls `create_generator` will be the same thread calling its `__next__`. For batched generation, this means that each thread will repeatedly poll for updates from the MLX worker thread. For sequential generation requests, each thread will be blocked on a lock that ensure sequential generation. For proper cleanup, ensure that the generator is closed or deleted when you no longer have a use for it.

`start` and `shutdown` functions were added to the model kits. This is most important for batched generation, where we need to manage a MLX worker thread. If you notice that your program (or tests) is not ending, ensure that the model kit is cleaned up with `mlx_engine.generate.shutdown()` or  `del model_kit`. For the most part this should be automatic, but python garbage collection is not necessarily guaranteed to run.

There are some race conditions that may prevent a request from being cancelled. For example, if the stop request is processed before the predict request. Or, if the stop request arrives after the prediction is done. In these cases, we print a logger warning. If we notice this warnings in practice, we should fix these race conditions.

## API changes

Added new `load_model()` parameters `max_seq_nums` and `seed`. The former controls how many decode sequences can be running in parallel, and the latter initializes the seed. Now that we are supporting batched generation, we cannot have each individual request reset the seed, as other sequences would be effected.

Added new `mlx_engine.generate` functions: `stop_generation(model_kit, request_id)` and `unload(model_kit)`. The former is the new API to cancel requests in progress. The latter function stops all generation and cleans up. For example, it shuts down the worker thread in the batched model kit.

Added new `request_id` parameter to `create_generator`. This is required so that we can cancel the correct request when required.

## Automatic backend detection

If the model supports images, or kv cache quantization is requested, or the model cache cannot be batch processed, use the (Vision)ModelKit backend. The existing backends will be removed (or dramatically refactored) down the line, once we bring up support for image model continuous batching.

It should be relatively simple to migrate kv cache quantization to the batched backend, once upstream support is added for it. `mlx_lm/server.py` contains an example of how to serve models that cannot be batched, and we can integrate similar code into our batched model kit.

## New demo

Added a batched_demo.py that uses the batched backend. This demo contains some functionality to easily display concurrent generation with a CLI command, without needing to set up a HTTP server (or similar). The original demo.py can be refactored (or removed) down the line once everything moves to batched generation. For now, preserve the original demo.py since we are still interested in observing the (Vision)ModelKit behavior.

## Refactors

Extracted GenerationResult, GenerationStopCondition, StopReason to a new file. No substantial changes were made to these

Moved StopPromptProcessing to prompt_progress_reporter.py, to be better shared across files

Added new BatchedMlxLmReporterAdapter for batched generation progress callbacks. The callbacks slightly differ from `mlx_lm.stream_generate`, so a new adapted reporter was added. This reporter does not throw `StopPromptProcessing` anymore. Once upstream supports cancelling prompt processing, we can start throwing this again.

The batched processer uses a default prompt prefill chunk size of `2048`. This means that some existing tests were edited to reduce the number of prompt processing callbacks. More test data was added in order to have more callbacks be generated now that the chunk size is larger.

## Known limitations

Speculative decoding is not implemented for batched generation. Hopefully this can be supported down the road soon. We could add functionality to unload the batched generator and load the sequential generator if we receive a speculative decoding load model API call, but this was elided from this PR in order to ship batched gen faster.

The batched backend doesn't report cached_tokens in prompt progress callbacks. I haven't invested time into writing the code to wire this information in. As a result, some test assertions are commented out.

Prompt processing cannot be interrupted in batched generation mode (yet). This is a high priority to fix, but this functionality is not yet exposed in `mlx_lm.BatchGeneration`.

Logging is not yet added for cases when requests are forcibly sequentialized.

## Codex review

Screenshot 2026-02-04 at 11 38 58 AM

## CLI demo

https://github.com/user-attachments/assets/89d94386-3ccf-456b-95fd-4fe7dd01aed8

# Authorship

The code in this PR was co-authored by @neilmehta24 and @christian-lms

### CODE REVIEW

File: mlx_engine/model_kit/batched_model_kit.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    Args:
+        model_path (Path): Path to the model directory containing model files.
+        max_kv_size (Optional[int]): Maximum size of the key-value cache.
+        max_seq_nums (Optional[int]): Maximum number of concurrent generation requests
+            that can be processed simultaneously.
+    """
+
+    model: nn.Module
+    tokenizer: TokenizerWrapper
+    model_type: str | None
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/model_kit/batched_model_kit.py]: `type | None` is syntactical sugar for `Optional[type]`. I prefer the former. The latter is required for Python < 3.10. The code uses both because I hand-write the former, and others (especially AI) like the latter. This code is running on 3.11 so we don't have a portability concern.

Both are fine, but we don't have our linter/formatter forcibly convert to one style (yet).

We have a couple inconsistencies like this, for example `list[type]` vs `List[type]`

File: mlx_engine/model_kit/batched_model_kit.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+                    if not found_request_id:
+                        logger.warning(f"Could not cancel {request_id=} (id not found)")
+                    continue
+
+                # Get cache
+                cache, rest = self._prompt_cache.fetch_nearest_cache(
+                    current_model_key, request.prompt_tokens
+                )
+                if cache is None:
+                    cache = make_prompt_cache(self.model)
```
[the remark this answers is not in the document]
Comment (neilmehta24) [reply mlx_engine/model_kit/batched_model_kit.py]: Good eye, removed this.
